Sources/Either/Either.swift

```diff
@@ -0,0 +1,210 @@
+  }
+}
+
+extension Either: Comparable where Left: Comparable, Right: Comparable {
+  public static func < (lhs: Either, rhs: Either) -> Bool {
+    switch (lhs, rhs) {
+    case let (.left(lhs), .left(rhs)):
+      return lhs < rhs
+    case let (.right(lhs), .right(rhs)):
+      return lhs < rhs
+    case (.left, .right):
+      return true
+    case (.right, .left):
+      return false
+    }
+  }
+}
+
+extension Either: Hashable where Left: Hashable, Right: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    self.do(
+      ifLeft: { hasher.combine($0) },
+      ifRight: { hasher.combine($0) }
+    )
+  }
+}
+
+public struct Errors: Error {
+  let errors: [Error]
+}
+
+extension Either: Decodable where Left: Decodable, Right: Decodable {
+  public init(from decoder: Decoder) throws {
+    do {
+      self = try .left(Left(from: decoder))
+    } catch let leftError {
+      do {
+        self = try .right(Right(from: decoder))
+      } catch let rightError {
+        throw DecodingError.typeMismatch(
+          Either.self,
+          .init(
+            codingPath: decoder.codingPath,
+            debugDescription: "Could not decode \(Left.self) or \(Right.self)",
+            underlyingError: Errors(errors: [leftError, rightError])
+          )
+        )
+      }
+    }
+  }
+}
+
+extension Either: Encodable where Left: Encodable, Right: Encodable {
+  public func encode(to encoder: Encoder) throws {
+    return try self.either(
+      ifLeft: { try $0.encode(to: encoder) },
+      ifRight: { try $0.encode(to: encoder) }
+    )
+  }
+}
+
+#if swift(>=5.0)
+extension Either where Left: Error {
+  public var asRightResult: Result<Right, Left> {
+    return self.either(ifLeft: Result.failure, ifRight: Result.success)
+  }
+}
+
+extension Either where Right: Error {
+  public var asLeftResult: Result<Left, Right> {
+    return self.either(ifLeft: Result.success, ifRight: Result.failure)
+  }
+}
+#endif
+
+extension Optional {
+  public func select<Left, Right>(
+    _ perform: ((Left) -> Right)?
+    ) -> Right?
+    where Wrapped == Either<Left, Right> {
+      return self.flatMap { e in
+        e.either(
+          ifLeft: { a in perform.map { f in f(a) } },
+          ifRight: { b in .some(b) }
+        )
+      }
+  }
+}
+
+extension Sequence {
+  public func lefts<Left, Right>() -> [Left] where Element == Either<Left, Right> {
+    return self.compactMap { $0.left }
+  }
+
+  public func rights<Left, Right>() -> [Right] where Element == Either<Left, Right> {
+    return self.compactMap { $0.right }
+  }
+
+  public func partitionMap<Left, Right>(_ transform: (Element) -> Either<Left, Right>)
+    -> ([Left], [Right]) {
+      return self.reduce(into: ([], [])) { result, element in
+        transform(element).do(
+          ifLeft: { result.0.append($0) },
+          ifRight: { result.1.append($0) }
+        )
+      }
+  }
+
+  public func partition(_ predicate: (Element) -> Bool) -> (pass: [Element], fail: [Element]) {
```

Comment:
This ships with the standard library in a strange way, so it would be good to only ship a "partition" story if it meshes well.